Initial release

pizzaandpineapples opened the pull request:
- Editor script to generate data from FMOD containing references to all banks, buses, VCAs, locales, events and parameters.
- FMODManager class is the entry point into using this library.
- EventsManager class is used to create emitters, play, pause, unpause, stop events and change parameters. Programmer sounds can also be played. 
- BanksManager class is used to load/unload banks, load/unload sample data and switch localization.
- The MixerManager is used to control the Buses and VCAs within FMOD.
- FMODCallBackHandler manages the internal states of each event. It can also handle looping audio. This callback is fired with every event that is played.
- FMODProgrammerSoundCallBackHandler manages the playback of externally loaded sounds and is required for dialogues and localized audio tables.
- CustomStudioEventEmitter is a modified FMOD emitter which handles initialization of event instances differently.
- Various data classes to hold data regarding FMOD components.

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- **New Features**
  - Introduced a comprehensive audio management system with initialization and event handling capabilities.
  - Added data classes for managing audio events and virtual control areas.

- **Documentation**
  - Updated unique identifiers for sample projects in the audio system package.

- **Chores**
  - Implemented automated workflows for testing, documentation generation, and Unity Package Manager (UPM) publishing.

- **Bug Fixes**
  - Addressed potential reference and dependency issues by updating asset GUIDs.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

pizzaandpineapples committed:
Added Editor Script to fetch all events from banks

pizzaandpineapples committed:
Modified editor script to fetch all parameters from events

pizzaandpineapples committed:
Created data classes for events

pizzaandpineapples committed:
Tested performance between event instances and emitters

pizzaandpineapples committed:
Expanded the AudioManager script

pizzaandpineapples committed:
Improving callbackhandler functionality

pizzaandpineapples committed:
Audio system can now load and unload banks

- There are errors in the editor script which will be fixed in the next commit.

pizzaandpineapples committed:
Bank load/unload now functions as desired

- Created a custom studio event emitter and a custom editor script for it.

pizzaandpineapples committed:
Cleaning up project files

pizzaandpineapples committed:
Added data classes for Bus and VCA

pizzaandpineapples committed:
AudioManager can be used to change FMOD Parameters

pizzaandpineapples committed:
Added callbackhandler for programmer sounds and started working on localization for audio

pizzaandpineapples committed:
Improving call back handler for dialogues

pizzaandpineapples committed:
Completed call back handler for dialogues

pizzaandpineapples committed:
Unit Testing

pizzaandpineapples committed:
Samples have been created and fixed some bugs

pizzaandpineapples committed:
FMODManager has been split into smaller components

pizzaandpineapples committed:
Refactoring project

pizzaandpineapples committed:
Fixed bugs and added documentation

pizzaandpineapples committed:
Update Sample.cs

pizzaandpineapples committed:
Fixed build errors

pizzaandpineapples committed:
Update README.md

pizzaandpineapples committed:
Updated README

Warhammer4000 commented on Assets/Packages/com.studio23.ss2.audiosystem.fmod/Runtime/Extensions/FMODProgrammerSoundCallBackHandler.cs:
```diff
@@ -0,0 +1,194 @@
+        /// <param name="eventData"></param>
+        /// <param name="key"></param>
+        /// <param name="assignCallBack"></param>
+        public static async void InitializeProgrammerCallback(FMODEmitterData eventData, string key, bool assignCallBack = false)
+        {
+            var data = await LoadExternalSound(eventData, key);
+            if (assignCallBack)
+            {
+                EVENT_CALLBACK eventCallback = new EVENT_CALLBACK(ProgrammerSoundCallbackHandler);
+                eventData.Emitter.EventInstance.setCallback(eventCallback);
+            }
+
+            GCHandle eventGcHandle = GCHandle.Alloc(data);
+            eventData.Emitter.EventInstance.setUserData(GCHandle.ToIntPtr(eventGcHandle));
+            eventData.Play();
+        }
```
This needs to be fixed.

Warhammer4000 commented on Assets/Packages/com.studio23.ss2.audiosystem.fmod/Editor/CreateFMODDataEditor.cs:
```diff
@@ -0,0 +1,396 @@
+using System.Linq;
+using UnityEditor;
+
+namespace Studio23.SS2.AudioSystem.fmod.Editor
+{
```
ignore this one

Warhammer4000 replied:
@coderabbitai please create a github issue as a follow up action

Warhammer4000 commented on Assets/Packages/com.studio23.ss2.audiosystem.fmod/Runtime/Core/EventsManager.cs:
```diff
@@ -0,0 +1,201 @@
+        /// <param name="referenceGameObject"></param>
+        /// <param name="emitter"></param>
+        /// <param name="stopModeType"></param>
+        public void PlayProgrammerSound(string key, FMODEventData eventData, GameObject referenceGameObject, CustomStudioEventEmitter emitter = null, STOP_MODE stopModeType = STOP_MODE.ALLOWFADEOUT)
+        {
+            var fetchData = EventEmitterExists(eventData, referenceGameObject);
+            if (fetchData != null)
+            {
+                FMODProgrammerSoundCallBackHandler.InitializeProgrammerCallback(fetchData, key);
+                return;
+            }
+            var newEmitter = new FMODEmitterData(eventData, referenceGameObject, emitter, stopModeType);
+            _emitterDataList.Add(newEmitter);
+            FMODProgrammerSoundCallBackHandler.InitializeProgrammerCallback(newEmitter, key, true);
+        }
```
InitializeProgrammerCallback can have a default bool parameter that's true if not passed. What do you think @coderabbitai

Warhammer4000 replied:
@coderabbitai would you be so kind as to create a pull request with the changes?

pizzaandpineapples replied:
The default value for ```assignCallBack``` parameter is false because we only need to set the callback once when creating an emitter. Every subsequent call requires it to be false.

Warhammer4000 commented on Assets/Packages/com.studio23.ss2.audiosystem.fmod/Runtime/Core/EventsManager.cs:
```diff
@@ -0,0 +1,201 @@
+        /// </summary>
+        /// <param name="eventData"></param>
+        /// <param name="referenceGameObject"></param>
+        public async void Play(FMODEventData eventData, GameObject referenceGameObject)
+        {
+            var fetchData = EventEmitterExists(eventData, referenceGameObject);
+            if (fetchData == null) return;
+            if (fetchData.EventState == FMODEventState.Playing) await fetchData.StopAsync(STOP_MODE.IMMEDIATE);
+            fetchData.Play();
+        }
```
need to  be fixed

pizzaandpineapples replied:
I need to await for the sound to be stopped before playing again.

Warhammer4000 commented on Assets/Packages/com.studio23.ss2.audiosystem.fmod/Runtime/Core/EventsManager.cs:
```diff
@@ -0,0 +1,201 @@
+            RuntimeManager.StudioSystem.setParameterByName(parameterName, parameterValue);
+        }
+
+        private FMODEmitterData EventEmitterExists(FMODEventData eventData, GameObject referenceGameObject)
+        {
+            return _emitterDataList.FirstOrDefault(x =>
+                x.BankName.Equals(eventData.BankName) && x.EventName.Equals(eventData.EventName) &&
+                x.ReferenceGameObject == referenceGameObject);
```
emitter data list probably should be a dictionary

pizzaandpineapples replied:
Yes please proceed with this modification.

pizzaandpineapples replied:
@coderabbitai please proceed with this modification.

Warhammer4000 commented on Assets/Packages/com.studio23.ss2.audiosystem.fmod/Runtime/Core/EventsManager.cs:
```diff
@@ -0,0 +1,201 @@
+                x.ReferenceGameObject == referenceGameObject);
+        }
+
+        internal async UniTask ClearEmitter(string bankPath)
+        {
+            for (int i = 0; i < FMODManager.Instance.EventsManager._emitterDataList.Count; i++)
+            {
+                FMODEmitterData emitter = FMODManager.Instance.EventsManager._emitterDataList[i];
+                if (emitter.BankName.Equals(bankPath))
+                {
+                    await emitter.ReleaseAsync();
+                    FMODManager.Instance.EventsManager._emitterDataList.Remove(emitter);
+                }
+            }
+        }
```
this is a good idea.

pizzaandpineapples replied:
The list has been converted a dictionary. @coderabbitai How do I iterate through it and remove the items?

Warhammer4000 commented on Assets/Packages/com.studio23.ss2.audiosystem.fmod/Runtime/Core/BanksManager.cs:
```diff
@@ -0,0 +1,109 @@
+        /// Unloads a Bank.
+        /// </summary>
+        /// <param name="bankName"></param>
+        public void UnloadBank(string bankName)
```
This is wrong. You already have a bank dictionary. get the bank from the string key and do the necessary operation in your remove bank method. What do you think @coderabbitai?

Warhammer4000 commented on Assets/Packages/com.studio23.ss2.audiosystem.fmod/Runtime/Core/BanksManager.cs:
```diff
@@ -0,0 +1,109 @@
+        /// Make sure to load the Sample Data for a Bank that has already been loaded.
+        /// </summary>
+        /// <param name="bankName"></param>
+        public void LoadBankSampleData(string bankName)
```
same you should be accessing bank like this
var bank=_banklist[bankName];
bank.loadSampleData();

@coderabbitai could you elaborate on this?

Warhammer4000 commented on Assets/Packages/com.studio23.ss2.audiosystem.fmod/Runtime/Core/MixerManager.cs:
```diff
@@ -0,0 +1,125 @@
+{
+    public class MixerManager
+    {
+        internal List<FMODBusData> _busDataList;
```
```suggestion
        internal List<FMODBusData> _busDataList;
```

This should be a dictionary of key value pairs please make methods defined bellow adapt to the changes @coderabbitai would you please review this MixerManager.cs file and suggest changes. for example the lists should be dictonaries

Warhammer4000 reviewed (changes requested):
Please make necessary changes

pizzaandpineapples committed:
Refactoring according to pull request

pizzaandpineapples commented on Assets/Packages/com.studio23.ss2.audiosystem.fmod/Runtime/Core/EventsManager.cs:
```diff
@@ -0,0 +1,206 @@
+        /// <param name="referenceGameObject"></param>
+        /// <param name="emitter"></param>
+        /// <param name="stopModeType"></param>
+        public async void PlayProgrammerSound(string key, FMODEventData eventData, GameObject referenceGameObject, CustomStudioEventEmitter emitter = null, STOP_MODE stopModeType = STOP_MODE.ALLOWFADEOUT)
+        {
+            var fetchData = EventEmitterExists(eventData, referenceGameObject);
+            if (fetchData != null)
+            {
+                await FMODProgrammerSoundCallBackHandler.InitializeProgrammerCallback(fetchData, key);
+                return;
+            }
+            var newEmitter = new FMODEmitterData(eventData, referenceGameObject, emitter, stopModeType);
+            _emitterDataList.Add(newEmitter.GetKey(), newEmitter);
+            await FMODProgrammerSoundCallBackHandler.InitializeProgrammerCallback(newEmitter, key, true);
+        }
```
```PlayProgrammerSound``` method should not be awaited on its own. ```async``` keyword is used so that we can use await inside.

pizzaandpineapples commented on Assets/Packages/com.studio23.ss2.audiosystem.fmod/Runtime/Extensions/FMODCallBackHandler.cs:
```diff
@@ -0,0 +1,89 @@
+        /// Initializes a default CallBack for all Event Instances.
+        /// </summary>
+        /// <param name="eventData"></param>
+        public static void InitializeCallBack(FMODEmitterData eventData)
+        {
+            EVENT_CALLBACK EventCallback = new EVENT_CALLBACK(EventCallbackHandler);
+            GCHandle EventGCHandle = GCHandle.Alloc(eventData);
+            eventData.Emitter.EventInstance.setUserData(GCHandle.ToIntPtr(EventGCHandle));
+            eventData.Emitter.EventInstance.setCallback(EventCallback);
+        }
```
Handles are freed when the Event Instances are destroyed.

pizzaandpineapples commented on Assets/Packages/com.studio23.ss2.audiosystem.fmod/Runtime/Core/MixerManager.cs:
```diff
@@ -0,0 +1,126 @@
+            _VCADataList = new Dictionary<string, FMODVCAData>();
+        }
+
+        private FMODBusData GetBus(string busName, float defaultVolume)
+        {
+            var newBus = new FMODBusData(busName, defaultVolume);
+            _busDataList.Add(newBus.BusName, newBus);
+            return newBus;
```
This method is private. It is never called directly. There is check whenever it is called through other methods.

pizzaandpineapples commented on Assets/Packages/com.studio23.ss2.audiosystem.fmod/Runtime/Extensions/FMODProgrammerSoundCallBackHandler.cs:
```diff
@@ -0,0 +1,194 @@
+                    return null;
+                }
+            }
+            else
+            {
+                // Load Sound Path
+                RESULT keyResult = RuntimeManager.StudioSystem.getSoundInfo(key, out soundInfo);
+                if (keyResult != RESULT.OK)
+                {
+                    Debug.LogError("Couldn't find dialogue with key: " + key);
+                    return null;
+                }
+
+                // Load Sound
+                RESULT soundResult = RuntimeManager.CoreSystem.createSound(soundInfo.name_or_data,
+                    soundMode | soundInfo.mode, ref soundInfo.exinfo, out sound);
+                if (soundResult != RESULT.OK)
+                {
+                    Debug.LogError("Couldn't load dialogue sound: " + key);
+                    return null;
+                }
+            }
+
+            //Wait to Load
+            int maxFrameWait = 120;
+            OPENSTATE openstate = OPENSTATE.BUFFERING;
+            while (openstate != OPENSTATE.READY)
+            {
+                await UniTask.Yield();
+                sound.getOpenState(out openstate, out uint percentbuffered, out bool starving, out bool diskbusy);
+                if (--maxFrameWait <= 0)
+                {
+                    sound.release();
+                    Debug.LogError("Failed to load sound " + key);
+                    return null;
+                }
+            }
+
+            return new SoundData(eventData, soundInfo, sound);
+        }
```
@Warhammer4000 should I make all ```Debug.LogError``` into ```Debug.LogWarning```?

Warhammer4000 replied:
Also @pizzaandpineapples  use path.combine instead of concat.

pizzaandpineapples committed:
Added a new method for sample data loading of single events

Warhammer4000 commented on Assets/Packages/com.studio23.ss2.audiosystem.fmod/Runtime/Core/BanksManager.cs:
```diff
@@ -0,0 +1,116 @@
+        public void UnloadAllBanks()
+        {
+            List<string> banksToRemove = _bankList.Keys.ToList();
+            foreach (var bank in banksToRemove)
```
you could just iterate through _bankList in a for each loop. it will give you a key value pair type. you could just pass the exact bank value in the unload method and work with it. 

@coderabbitai could you provide a sample for what I just said?

Warhammer4000 commented on Assets/Packages/com.studio23.ss2.audiosystem.fmod/Runtime/Core/FMODManager.cs:
```diff
@@ -0,0 +1,69 @@
+        public bool InitializeOnStart;
+        
+        private static FMODManager _instance;
+        public static FMODManager Instance
```
this is probably a bad idea as this manager should persist through the games lifecycle. So it should have a dont destroy on load method.

Warhammer4000 commented on Assets/Packages/com.studio23.ss2.audiosystem.fmod/Runtime/Data/FMODBusData.cs:
```diff
@@ -0,0 +1,85 @@
+{
+    public class FMODBusData
+    {
+        public Bus Bus;
+        public string BusName;
+        public float DefaultVolume;
+        public float CurrentVolume;
+        public bool IsPaused;
+        public bool IsMuted;
```
shouldn't they all be internal?

Warhammer4000 commented on Assets/Packages/com.studio23.ss2.audiosystem.fmod/Runtime/Data/FMODEmitterData.cs:
```diff
@@ -0,0 +1,150 @@
+    [System.Serializable]
+    public class FMODEmitterData
+    {
+        public string BankName;
+        public string EventName;
+        public string EventGUID;
+        public GameObject ReferenceGameObject;
+        public CustomStudioEventEmitter Emitter;
+        public FMODEventState EventState = FMODEventState.Stopped;
+        public STOP_MODE StopModeType;
+        public EVENT_CALLBACK_TYPE CurrentCallbackType;
```
shouldn't they all be internal?

Warhammer4000 reviewed (changes requested):
Please make the ncessary changes & add the CI pipleine files for 
1) Test Runner
2) DocGen
3)UpmRelease

pizzaandpineapples committed:
Further refactoring and bug fixes

pizzaandpineapples committed:
Refactoring for upm release

- Bug fixing.
- Samples folder updated.
- CI pipeline files for Test Runner, DocGen, UpmRelease added.

pizzaandpineapples committed:
Updated package.json and play mode tests

pizzaandpineapples committed:
Update testrunner.yml

pizzaandpineapples committed:
Update testrunner.yml

N-Evan commented:
@coderabbitai review

pizzaandpineapples committed:
Update PlayModeTests.cs

N-Evan reviewed (commented):
Looks good. Proceed.

pizzaandpineapples committed:
Update PlayModeTests.cs

pizzaandpineapples committed:
Update PlayModeTests.cs

Warhammer4000 commented on .github/workflows/docgen.yml:
```diff
@@ -0,0 +1,42 @@
+        with:
+          submodules: true
+      
+      - name: Find and copy specific files
+        run: |
+            # Look for LICENSE.md, README.md, and CHANGELOG.md in subdirectories
+            find . -name LICENSE.md -exec cp {} . \;
+            find . -name README.md -exec cp {} . \;
+            find . -name CHANGELOG.md -exec cp {} . \;
+
+            # Check if a "Screenshots" folder exists and copy it to the root
+            if [ -d "Screenshots" ]; then
+              cp -r Screenshots .
+            fi
```
what is your recommendation @coderabbitai?

Warhammer4000 commented on .github/workflows/upmpublishing.yml:
```diff
@@ -0,0 +1,48 @@
+    - name: Push Branch 
+      run: |
+        git checkout upm
+        if [[ -d "Samples" ]]; then
+          git mv Samples Samples~
+          rm -f Samples.meta
+          git config --global user.name 'github-bot'
+          git config --global user.email 'github-bot@users.noreply.github.com'
+          git commit -am "fix: Samples => Samples~"
```
this is intended. samples folder shouldn't exist when we publish package. But we still want to be able to import it via some other mechanism. hence this design
